presto-orc/src/main/java/com/facebook/presto/orc/writer/FlatMapColumnWriter.java

```diff
@@ -0,0 +1,289 @@
+    private final ColumnWriterFactory valueWriterFactory;
+    private final int stringStatisticsLimitInBytes;
+
+    private final LongOutputStream lengthStream;
+    private final PresentOutputStream presentStream;
+    private List<FlatMapKeyNodeWriter> keyNodes;
+    /* Block Writer for Numeric or String Keyed Types */
+    private final FlatMapBlockWriter blockWriter;
+    private StatisticsBuilder statisticsBuilder;
+    private final Supplier<StatisticsBuilder> statisticsBuilderSupplier;
+
+    private final List<ColumnStatistics> rowGroupColumnStatistics = new ArrayList<>();
+    private long columnStatisticsRetainedSizeInBytes;
+
+    private int nonNullValueCount;
+    private boolean closed;
+
+    private StringStatisticsBuilder newStringStatisticsBuilder()
+    {
+        return new StringStatisticsBuilder(stringStatisticsLimitInBytes);
+    }
+
+    public FlatMapColumnWriter(
+            int column,
+            int sequence,
+            ColumnWriterOptions columnWriterOptions,
+            Optional<DwrfDataEncryptor> dwrfEncryptor,
+            OrcEncoding orcEncoding,
+            MetadataWriter metadataWriter,
+            Type type,
+            List<OrcType> orcTypes,
+            ColumnWriterFactory valueWriterfactory)
+    {
+        checkArgument(column >= 0, "column is negative");
+        checkArgument(sequence >= 0, "sequence is negative");
+
+        this.column = column;
+        this.sequence = sequence;
+        this.keyColumnId = orcTypes.get(column).getFieldTypeIndex(0);
+        this.valueColumnId = orcTypes.get(column).getFieldTypeIndex(1);
+
+        this.type = (MapType) type;
+        this.keyType = type.getTypeParameters().get(0);
+        this.valueType = type.getTypeParameters().get(1);
+
+        this.keyOrcType = orcTypes.get(keyColumnId);
+
+        this.compressed = columnWriterOptions.getCompressionKind() != NONE;
+
+        this.columnEncoding = new ColumnEncoding(orcEncoding == DWRF ? DWRF_MAP_FLAT : DIRECT_V2, 0);
+        this.keyColumnEncoding = new ColumnEncoding(orcEncoding == DWRF ? DIRECT : DIRECT_V2, 0);
+        this.valueColumnEncoding = new ColumnEncoding(orcEncoding == DWRF ? DIRECT : DIRECT_V2, 0);
+
+        this.dwrfEncryptor = dwrfEncryptor;
+        this.metadataWriter = new CompressedMetadataWriter(metadataWriter, columnWriterOptions, dwrfEncryptor);
+        this.inputMetadataWriter = metadataWriter;
+
+        this.valueWriterFactory = valueWriterfactory;
+        this.stringStatisticsLimitInBytes = toIntExact(columnWriterOptions.getStringStatisticsLimit().toBytes());
+
+
+        this.lengthStream = createLengthOutputStream(columnWriterOptions, dwrfEncryptor, orcEncoding);
+        this.presentStream = new PresentOutputStream(columnWriterOptions, dwrfEncryptor);
+        this.keyNodes = new ArrayList<>();
+
+        switch (keyOrcType.getOrcTypeKind()) {
+            case CHAR:
+                checkArgument(orcEncoding != DWRF, "DWRF does not support %s type", type);
```

Comment:
FlatMap is only supported for DWRF and this condition is not DWRF, so char can be removed.